### PR TITLE
Specialize serialization for ArrayVectors

### PR DESCRIPTION
**Pull request by dnhatn**
Currently, we serialize blocks and vectors value by value, employing a simple yet effective approach. However, there are specific cases where we can enhance performance by serializing the underlying structure instead:

1. Serializing BytesRefArray of a BytesRefArrayVector.
2. Serializing the `firstValueIndexes`, `nullsMask`, and the underlying vector of an ArrayBlock instead of rebuilding the block from values.
3. Serializing BigArrayBlock.

This PR addresses the first bullet point and lays the groundwork for implementing the second.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BytesRefArrayVector.java [hunk 1]
```diff
@@ -30,6 +34,25 @@ final class BytesRefArrayVector extends AbstractVector implements BytesRefVector
         this.values = values;
     }
 
+    static BytesRefArrayVector readArrayVector(int positions, StreamInput in, BlockFactory blockFactory) throws IOException {
+        final BytesRefArray values = new BytesRefArray(in, blockFactory.bigArrays());
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BytesRefArrayVector.java:38]: This is the main improvement in this PR. The rest is for a follow-up where we will implement specialized serialization for ArrayBlock.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/Vector.java [hunk 2]
```diff
@@ -75,4 +75,8 @@ interface Builder extends Releasable {
      * Whether this vector was released
      */
     boolean isReleased();
+
+    byte SERIALIZE_VECTOR_VALUES = 0;
+    byte SERIALIZE_VECTOR_CONSTANT = 1;
+    byte SERIALIZE_VECTOR_ARRAY = 2;
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/Vector.java:81]: Worth some comments I think. Worth explaining that we use 0 and 1 for backwards compatibility.

**Review by nik9000 — APPROVED** [review]

**Comment by dnhatn** [comment]
Thanks Nik!

**Review by ChrisHegarty — COMMENTED** [review]
I love it. And how we can evolve the serial form in a BWC way through the initial Boolean/byte. This was always my hope - nice to see it work in reality. Belated LGTM

**Pull request merged**